src/main/scala/chiseltest/formal/backends/smt/SMTExprEval.scala

```diff
@@ -0,0 +1,163 @@
+  private def doBVEqual(a:  BigInt, b:     BigInt): BigInt = bool(a == b)
+  private def doBVCompare(op: Compare.Value, a: BigInt, b: BigInt, width: Int, signed: Boolean): BigInt = {
+    if (a == b && op == Compare.GreaterEqual) return bool(true)
+
+    // only checking for Greater than
+    if (signed) {
+      (isPositive(a, width), isPositive(b, width)) match {
+        case (true, true)  => doBVCompare(op, a, b, width - 1, signed = false)
+        case (true, false) =>
+          // a < 0 && b >= 0 => a can never be greater or equal to b
+          bool(false)
+        case (false, true) =>
+          // a >= 0 && b < 0 => a is greater than b
+          bool(true)
+        case (false, false) => doBVCompare(op, doBVNegate(b, width), doBVNegate(a, width), width - 1, signed = false)
+      }
+    } else {
+      bool(a > b)
+    }
+
+  }
+  private def doBVOp(op: Op.Value, a: BigInt, b: BigInt, width: Int): BigInt = op match {
+    case Op.Xor       => a ^ b
+    case Op.ShiftLeft => (a << b.toInt) & mask(width)
+    case Op.ArithmeticShiftRight =>
+      val by = b.toInt
+      if (isPositive(a, width)) { a >> by }
+      else if (by >= width) { mask(width) }
+      else {
+        val msb = mask(by) << (width - by)
+        (a >> by) | msb
+      }
+    case Op.ShiftRight => a >> b.toInt
+    case Op.Add        => (a + b) & mask(width)
+    case Op.Mul        => (a * b) & mask(width)
+    case Op.Sub        => sub(a, b, width)
+    case other         => throw new NotImplementedError(other.toString)
+  }
+  private def doBVConcat(a: BigInt, b: BigInt, bWidth: Int): BigInt = (a << bWidth) | b
+
+  // helper functions
+  private def sub(a:            BigInt, b: BigInt, width: Int): BigInt = (a + flipBits(b, width) + 1) & mask(width)
+  private def mask(width:       Int): BigInt = (BigInt(1) << width) - 1
+  private def isPositive(value: BigInt, w: Int) = (value & mask(w - 1)) == value
+  private def isNegative(value: BigInt, w: Int) = !isPositive(value, w)
+  private def flipBits(value:   BigInt, w: Int) = ~value & mask(w)
+  private def bool(b:           Boolean): BigInt = if (b) BigInt(1) else BigInt(0)
+}
+
+private[chiseltest] case class LocalEvalCtx(bv: Map[String, BigInt], array: Map[String, ArrayValue] = Map())
+    extends SMTEvalCtx {
+  override def getBVSymbol(name: String): BigInt = bv.getOrElse(
+    name, {
+      variables.find(_._1 == name).map(_._2).getOrElse(throw new RuntimeException(s"Unknown symbol $name!"))
+    }
+  )
+  override def getArraySymbol(name: String) = array(name)
+
+  private val variables = scala.collection.mutable.Stack[(String, BigInt)]()
+  override def startVariableScope(name: String, value: BigInt): Unit = {
+    variables.push((name, value))
+  }
+  override def endVariableScope(name: String): Unit = {
+    val old = variables.pop()
+    assert(old._1 == name)
+  }
+  override def constArray(indexWidth: Int, value: BigInt) = ???
```

Comment:
Maybe a comment with a Todo entry here

Reply:
To be honest, I do not actually know when this would be used.
For now it is at least not a regression over the old `maltese-smt` code.